common/src/main/java/com/graphicsfuzz/common/util/FuzzyImageComparison.java

```diff
@@ -0,0 +1,518 @@
+
+    public static int EXIT_STATUS_SIMILAR = 0;
+    public static int EXIT_STATUS_DIFFERENT = 1;
+
+    public MainResult(
+        boolean areImagesDifferent,
+        int exitStatus,
+        List<ThresholdConfiguration> configurations) {
+      this.areImagesDifferent = areImagesDifferent;
+      this.exitStatus = exitStatus;
+      this.configurations = configurations;
+    }
+  }
+
+  private static boolean arePixelsSimilarComponentThreshold(
+      int colorLeft,
+      int colorRight,
+      int componentThreshold) {
+
+    for (int i = 0; i < 32; i += 8) {
+      int left = ImageColorComponents.getComponent(colorLeft, i);
+      int right = ImageColorComponents.getComponent(colorRight, i);
+      assert left >= 0 && left <= 0xff;
+      assert right >= 0 && right <= 0xff;
+      if (Math.abs(left - right) > componentThreshold) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private static boolean doesSimilarNearPixelExist(
+      final int[] colorsLeft,
+      final int[] colorsRight,
+      final int width,
+      final int height,
+      final int componentThreshold,
+      final int distanceThreshold,
+      final int middleX,
+      final int middleY) {
+
+    final int middlePos = middleY * height + middleX;
+
+    final int ystart = Math.max(0, middleY - distanceThreshold);
+    final int xstart = Math.max(0, middleX - distanceThreshold);
+    final int yend = Math.min(height, middleY + distanceThreshold);
+    final int xend = Math.min(width, middleX + distanceThreshold);
+
+    for (int y = ystart; y < yend; ++y) {
+      for (int x = xstart; x < xend; ++x) {
+        int pos = y * height + x;
+        if (arePixelsSimilarComponentThreshold(
+            colorsLeft[middlePos],
+            colorsRight[pos],
+            componentThreshold)) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  /**
+   * @param badPixels will contain a non-zero value at every bad pixel found
+   * @return the number of bad pixels found according to the thresholds provided
+   */
+  private static int compareImageColors(
+      int[] colorsLeft,
+      int[] colorsRight,
+      int width,
+      int height,
+      int componentThreshold,
+      int distanceThreshold,
+      int[] badPixels) {
+
+    assert colorsLeft.length == colorsRight.length;
+
+    int numBad = 0;
+
+    for (int y = 0; y < height; ++y) {
+      for (int x = 0; x < width; ++x) {
+
+        // Given pixel in image A, find similar, nearby pixel in image B, and vice-versa.
+        // If either fails (hence || below), then the pixel coordinate is bad.
+        // See comments at top of file for justification.
+        if (!doesSimilarNearPixelExist(
+            colorsLeft,
+            colorsRight,
+            width,
+            height,
+            componentThreshold,
+            distanceThreshold,
+            x,
+            y)
+            ||
+            !doesSimilarNearPixelExist(
```

Comment:
The indentation of arguments of this function call is inconsistent with the indentation right above for the previous call of doesSimilarNearPixelExist()

Reply:
Thanks. Done.